packages/jest-snapshot/src/getMatchers.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+'use strict';
+
+const common = require('./common');
+const fs = require('fs');
+const path = require('path');
+const serializer = require('./serializer');
+const Snapshot = require('./Snapshot');
+
+const testsByPath =  {};
+const getNextIndexFor = (test, filePath) => {
+  const tests = testsByPath[filePath] || (testsByPath[filePath] = {});
+  if (test in tests) {
+    return ++tests[test];
+  }
+  return tests[test] = 0;
+};
+
+module.exports = (filePath, options) => ({
+  toMatchSnapshot: (util, customEquality) => {
+    return {
+      compare(actual, expected) {
+        // Retrieving last test asap to avoid race condition
+        const lastTest = common.getLastTest(filePath);
```

Comment:
Jest is the only process keeping the real state and spawning processes, initially in order to be sure which file I was working on I was reading it in the stack call by creating an Error and get the filepath there.
After writing some stress tests I was always getting some race condition, I rewrote the code in this way and there's no more race condition possible, as jest keeps a state for each file and the file can only run a single test at a time as Jasmine runs always one Spec at a time (as it uses setTimeout https://github.com/jasmine/jasmine/blob/master/src/core/QueueRunner.js#L21 which run on a single thread), it's possible to pass a custom TreeProcessor but we are currently not doing it, I suppose that in that case it's possible to ovveride this by really messing up stuff :).

I checked and by passing down the filepath from jest I was correctly getting the same path as it was in the stack trace 100% of the time even under stress.